server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesMethod.java

```diff
@@ -0,0 +1,137 @@
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+public class SetMessagesMethod<Id extends MailboxId> implements Method {
+
+    private static final int LIMIT_BY_ONE = 1;
+    private static final Logger LOGGER = LoggerFactory.getLogger(SetMessagesMethod.class);
+    private static final Method.Request.Name METHOD_NAME = Method.Request.name("setMessages");
+    private static final Method.Response.Name RESPONSE_NAME = Method.Response.name("messagesSet");
+
+    private final MailboxMapperFactory<Id> mailboxMapperFactory;
+    private final MailboxSessionMapperFactory<Id> mailboxSessionMapperFactory;
+
+    @Inject
+    @VisibleForTesting public SetMessagesMethod(MailboxMapperFactory<Id> mailboxMapperFactory, MailboxSessionMapperFactory<Id> mailboxSessionMapperFactory) {
+        this.mailboxMapperFactory = mailboxMapperFactory;
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+    }
+
+    @Override
+    public Method.Request.Name requestHandled() {
+        return METHOD_NAME;
+    }
+
+    @Override
+    public Class<? extends JmapRequest> requestType() {
+        return SetMessagesRequest.class;
+    }
+
+    public Stream<JmapResponse> process(JmapRequest request, ClientId clientId, MailboxSession mailboxSession) {
+        Preconditions.checkArgument(request instanceof SetMessagesRequest);
+        try {
+            return Stream.of(
+                    JmapResponse.builder().clientId(clientId)
+                    .response(setMessagesResponse((SetMessagesRequest) request, mailboxSession))
+                    .responseName(RESPONSE_NAME)
+                    .build());
+        } catch (MailboxException e) {
+            return Stream.of(
+                    JmapResponse.builder().clientId(clientId)
+                    .error()
+                    .responseName(RESPONSE_NAME)
+                    .build());
+        }
+    }
+
+    private SetMessagesResponse setMessagesResponse(SetMessagesRequest request, MailboxSession mailboxSession) throws MailboxException {
+        SetMessagesResponse.Builder builder = SetMessagesResponse.builder();
+        processDestroy(request.getDestroy(), mailboxSession, builder);
+        return builder.build();
+    }
+
+    private void processDestroy(List<MessageId> messageIds, MailboxSession mailboxSession, SetMessagesResponse.Builder builder) throws MailboxException {
+        MessageMapper<Id> messageMapper = mailboxSessionMapperFactory.createMessageMapper(mailboxSession);
+        Consumer<? super MessageId> delete = delete(messageMapper, mailboxSession, builder);
+
+        messageIds.stream()
+            .forEach(delete);
+    }
+
+    private Consumer<? super MessageId> delete(MessageMapper<Id> messageMapper, MailboxSession mailboxSession, Builder builder) {
+        return (messageId) -> {
+            try {
+                Mailbox<Id> mailbox = mailboxMapperFactory
```

Comment:
That's not the same process in IMAP: 
- STORE command to set Deleted flag
- then EXPUNGE command is called

I will investigate where this delete method is called.